### DISCUSSION TITLE
Tips for getting live preview working in Xcode 12.5

### DISCUSSION THREAD

**Discussion by surajbarkale** [pointfreeco/swift-composable-architecture#581] [General]
Hi,

I have an app structured like _isowords_ and I had a lot of trouble getting SwiftUI previews working. The best workaround was to add all packages that contain views (not just AppFeature) as dependencies for the app target in Xcode.

It appears that the SwiftUI previews only work for the App project and it's direct dependencies.

I have lodged FB9118422 with Apple where I kept on getting `failedToGenerateThunkInfo` error on previews.

Regards,
Suraj

**Comment by mbrandonw** [comment]
Seems like a bug in Xcode 12.5 😕. Some of our previews work in isowords and others do not. No idea why.

**Comment by surajbarkale** [comment]
Yes I just ended up adding the UI features as dependencies to project to get them working:
![image](https://user-images.githubusercontent.com/1241310/120409004-7413de00-c393-11eb-8214-cad680cd8784.png)

**Reply by mbrandonw** [reply]
Yeah, but that's a real bummer because now you are forced to build the entire application just to run a preview. This is a pretty serious Xcode bug. Hope it gets fixed soon.

**Reply by maximkrouk** [reply]
Does not work with my project 🥲

**Comment by ferologics** [comment]
Submitted feedback `FB9186658` for the Preview Crash that starts occurring when I try to use a `ViewStore` my previews.

**Reply by andreweades** [reply]
I submitted feedback too. 
The Apple engineer who is looking into it gave a "pro tip" on Twitter in response:

"Pro-tip: Xcode looks at the dependency graph to determine which app to use for hosting previews. If you're working in a framework/package, and you have an app target that depends on it, and the app is in the scheme, we'll launch the app to host previews to get entitlements, etc."

So if you make a mini-app for each feature or even every individual view, you can then select the scheme for the app and use previews in the package as it will pick the app to use as the preview host.
This means you can get some way to how *we* think it should work. The app does not have to have anything in it as far as I can tell. You can just leave the "Hello, World" in so long as you set the dependency to the package that contains the views.

**Reply by mbrandonw** [reply]
I really hope Apple considers this a bug and not how previews are supposed to be used. It's counterintuitive and a lot of work to maintain. Kinda kills the whole flow of using SPM to keep things modularized in a lightweight way.

**Reply by andreweades** [reply]
I agree. it seems like a backwards step to me.

**Reply by clayellis-cricut** [reply]
For those looking for tweet mentioned above:

https://mobile.twitter.com/kbcathey/status/1405162790298341382

**Comment by tgrapperon** [comment]
I too have difficulties with Previews.

One thing that is working better for me is to open `Package.swift` instead of the workspace. I've seen the tweet explaining that Xcode tries to find an app in the workspace to host your preview, and I wondered what would happen if there is no app to begin with.

You can't launch an app nor have local overrides for your dependencies like this, but it seems that Xcode is less confused, and it's working quite reliably for me as long as the current scheme is building the file I'm previewing.

This is not optimal though, but it seems simpler than trying to delete build product, or derived data, or restarting Xcode without success.
